src/java/org/apache/cassandra/gms/Gossiper.java

```diff
@@ -2343,22 +2398,36 @@ public boolean waitForSchemaAgreement(long maxWait, TimeUnit unit, BooleanSuppli
      */
     public boolean hasMajorVersion3Nodes()
     {
-        return isUpgradingFromVersionLowerThan(CassandraVersion.CASSANDRA_4_0) || // this is quite obvious
-               // however if we discovered only nodes at current version so far (in particular only this node),
-               // but still there are nodes with unknown version, we also want to report that the cluster may have nodes at 3.x
-               upgradeInProgressPossible && !isUpgradingFromVersionLowerThan(SystemKeyspace.CURRENT_VERSION.familyLowerBound.get());
+        return isUpgradingFromVersionLowerThan(CassandraVersion.CASSANDRA_4_0);
     }
 
     /**
      * Returns {@code true} if there are nodes on version lower than the provided version
      */
     public boolean isUpgradingFromVersionLowerThan(CassandraVersion referenceVersion)
     {
-        CassandraVersion v = upgradeFromVersionMemoized.get();
-        if (CassandraVersion.NULL_VERSION.equals(v) && scheduledGossipTask == null)
-            return false;
+        return getMinVersion().compareTo(referenceVersion) < 0;
+    }
 
-        return v != null && v.compareTo(referenceVersion) < 0;
+    /**
+     * This is a safe way to get the version we are upgrading from. It will return the current version if the cluster
+     * is not in the upgrade state. If the cluster is in upgrade state, it will return NULL_VERSION, if there is no
+     * information about the other nodes yet. Otherwise, it will just return the minimum cluster version.
+     */
+    public CassandraVersion getMinVersion()
+    {
+        if (!clusterVersionProvider.isUpgradeInProgress())
+            return clusterVersionProvider.getMinClusterVersion();
+
+        CassandraVersion v = clusterVersionProvider.getMinClusterVersion();
+        assert v != null : "API contract violation: cluster version provider implementation should never return null";
```

Comment:
`clusterVersionProvider.getMinClusterVersion()` may be returned 3 lines above, so the assert should be moved up, before that first use